source/include/util/cs_CuckooFilter.h

```diff
@@ -0,0 +1,246 @@
+ * and many implementation details have changed e.g. to fix implicit narrowing/widening of integers,
+ * large recursion and type punning in allocations.
+ *
+ * Those changes have been made by the Crownstone Team and fall under the project license mentioned above.
+ *
+ * The original code and the extent to which is required by applicable law is left under its original license
+ * included below and is attributed to the original author Jonah H. Harris <jonah.harris@gmail.com>.
+ *
+ * The MIT License
+ *
+ * Copyright (c) 2015 Jonah H. Harris
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+
+#pragma once
+
+#include <protocol/cs_CuckooFilterStructs.h>
+
+/**
+ * A CuckooFilter is a probabilistic datatype is made for approximate membership queries.
+ *
+ * This implementation consists of the POD data structure `cuckoo_filter_data_t` and a
+ * 'view class' `Cuckoofilter`, which is used to manipulate the data struct.
+ *
+ * The cuckoo_filter_data_t struct is not aware of its size. That must be managed by the owning
+ * scope, which can query the required size in bytes for the given amount of fingerprints through
+ * the CuckooFilter class.
+ *
+ * The 'elements' that one can `add` to a `CuckooFilter` are of type `cuckoo_key_t` which are
+ * simply sequences of bytes. To check if an element is contained in a filter, refer to the
+ * `contains` methods.
+ *
+ * CuckooFilter data contents are stored in an array of fingerprints which is segmented into
+ * 'buckets'. These buckets are used for indexing, and although for an element, the bucket index and
+ * fingerprint are deterministic, the index inside the bucket is dependent on order of insertion.
+ *
+ * More details can be found in `https://www.cs.cmu.edu/~dga/papers/cuckoo-conext2014.pdf`.
+ * "Cuckoo Filter: Better Than Bloom" by Bin Fan, Dave Andersen, and Michael Kaminsky
+ */
+class CuckooFilter {
```

Comment:
I always prefer having the public API on top, as that's what users will be looking for. Right now, the public functions are the least documented ones.

Instead of returning a bool, consider returning an error code, it's handy to be more specific about the error, and you can use it as result for control commands.

Reply:
This was legacy: when the cuckoo_data_t struct was an inner class, it needed to be defined before the methods and a forward declaration seemed inappropriate. I like having the public stuff on top too :-) 

*fixed